### PR TITLE
Updates to API endpoints for pin and unpin

### PR DESCRIPTION
**Pull request by zbiljic**
Adds new API endpoints for "pinning" chunks fro different types of content.

The new endpoints are following:
* `/pinning/bytes`
* `/pinning/files`
* `/pinning/bzz`

Each of these endpoints have `POST` and `DELETE`, to pin some structure chunks, and delete them. Internally on each endpoint type is checked for the supplied reference, and if it of expected type, traversal of all chunks is performed for the type, and each found chunk is pinned.
No separate internal index was created for these different structures, they are all added to existing `pinIndex`, which stores chunk hash to number of times it was pinned.

To access all the pinned chunks, one would use already existing `/pinning/chunks` endpoint, using `GET` method.

### CODE REVIEW

**Review by janos — APPROVED** [review]
👍

File: openapi/Swarm.yaml [hunk 1]
```diff
@@ -557,6 +557,147 @@ paths:
         default:
           description: Default response
 
+  '/pinning/bytes/{address}':
```
Comment (zelig) [inline openapi/Swarm.yaml:560]: i would really prefer just `pin` and not `pinning`

File: pkg/api/pinning_bytes.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/traversal"
+	"github.com/gorilla/mux"
+)
+
+// pinBytesUploaded is used to pin an already uploaded content.
+func (s *server) pinBytesUploaded(w http.ResponseWriter, r *http.Request) {
```
Comment (zelig) [inline pkg/api/pinning_bytes.go:19]: `pinBytes`?

File: pkg/api/pinning_bytes.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		s.Logger.Error("pin bytes: cannot pin")
+		jsonhttp.InternalServerError(w, "cannot pin")
+		return
+	}
+
+	jsonhttp.OK(w, nil)
+}
+
+// pinBytesRemovePinned removes pinning from content.
+func (s *server) pinBytesRemovePinned(w http.ResponseWriter, r *http.Request) {
```
Comment (zelig) [inline pkg/api/pinning_bytes.go:73]: `unpinBytes`?

File: pkg/api/pinning_bzz.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	}
+
+	if !has {
+		jsonhttp.NotFound(w, nil)
+		return
+	}
+
+	ctx := r.Context()
+
+	chunkAddressFn := func(address swarm.Address) (stop bool) {
```
Comment (zelig) [inline pkg/api/pinning_bzz.go:43]: this function should be the same across all API calls maybe and taken from the `pin` package

File: pkg/api/pinning_bzz.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/traversal"
+	"github.com/gorilla/mux"
+)
+
+// pinBzzUploaded is used to pin an already uploaded content.
+func (s *server) pinBzzUploaded(w http.ResponseWriter, r *http.Request) {
```
Comment (zelig) [inline pkg/api/pinning_bzz.go:19]: `pinBzz`?

File: pkg/api/pinning_bzz.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+		s.Logger.Error("pin bzz: cannot pin")
+		jsonhttp.InternalServerError(w, "cannot pin")
+		return
+	}
+
+	jsonhttp.OK(w, nil)
+}
+
+// pinBzzRemovePinned removes pinning from content.
+func (s *server) pinBzzRemovePinned(w http.ResponseWriter, r *http.Request) {
```
Comment (zelig) [inline pkg/api/pinning_bzz.go:73]: `unpinBzz`

File: pkg/api/pinning_files.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/traversal"
+	"github.com/gorilla/mux"
+)
+
+// pinFilesUploaded is used to pin an already uploaded content.
+func (s *server) pinFilesUploaded(w http.ResponseWriter, r *http.Request) {
```
Comment (zelig) [inline pkg/api/pinning_files.go:19]: `pinFile`?

File: pkg/api/pinning_files.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+		s.Logger.Error("pin files: cannot pin")
+		jsonhttp.InternalServerError(w, "cannot pin")
+		return
+	}
+
+	jsonhttp.OK(w, nil)
+}
+
+// pinFilesRemovePinned removes pinning from content.
+func (s *server) pinFilesRemovePinned(w http.ResponseWriter, r *http.Request) {
```
Comment (zelig) [inline pkg/api/pinning_files.go:73]: `unpinFile`?

**Review by zelig — APPROVED** [review]
very nice job @zbiljic 
I would prefer renaming 
- the API endpoint from `pinning` to `pin` 
- the filenames containing `pinning` to `pin`
- the service API calls from `pinBytesUploaded/pinBytesRemovePinned` to `pinBytes/unpinBytes`, etc
- the chunkAddressFn ought to be a general chunk `pin/unpin` function and defined within the `pin` package

**Comment by zbiljic** [comment]
@zelig Sure thing, I will rename everything.
Initially, I have changed `pinning` API endpoint to `pin`, but that was when I was thinking about having single endpoint, where reference would be detected. Now that we have multiple types, we will lose ability to add separate `pin` API in the future.

**Review by zelig — APPROVED** [review]

**Review by janos — APPROVED** [review]

**Comment by ldeffenb** [comment]
The API docs at https://docs.ethswarm.org/api/#tag/Chunk-pinning do not reflect this change.  Is there a newer API doc that I should be referencing for the upcoming release?

**Pull request merged**